src/Retrier.php

```diff
@@ -0,0 +1,172 @@
+<?php
+/*
+ * Copyright 2024 Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+namespace Google\ApiCore;
+
+use Google\ApiCore\ApiException;
+use Google\ApiCore\ApiStatus;
+use Google\ApiCore\RetrySettings;
+
+/**
+ * Retrier functionality.
+ */
+class Retrier
+{
+    private RetrySettings $retrySettings;
+
+    public function __construct(
+        RetrySettings $retrySettings
+    ) {
+        $this->retrySettings = $retrySettings;
+    }
+
+    /**
+     * @return RetrySettings
+     */
+    public function getRetrySettings()
+    {
+        return $this->retrySettings;
+    }
+
+    /**
+     * Execute the callable with the retry logic.
+     *
+     * @param callable $call
+     * @param array $options
+     *
+     * @return mixed
+     * @throws \Exception
+     */
+    public function execute(callable $call, array $options)
```

Comment:
This is certainly something new, but do we want to add the ability to change the `$call` or `$options` after a failed attempt.

We had this ability in Core where we allowed changing of headers using `retryListeners`.
Although we don't need this functionality right now, but I just wanted to initiate a discussion on this.